### PR TITLE
operator/ciliumidentity: Add CID time trackers

### PR DESCRIPTION
**Pull request by ovidiutirla**
Add `EnqueueTimeTracker` and `CIDDeletionTracker` structures to manage enqueuing times and track CID deletion marks. This feature will be used by Operator Managing CIDs. 

`EnqueueTimeTracker` is only used for metrics to meter the duration from enqueuing the reconciliation until the CID reconciliation is completed. 
`CIDDeletionTracker` is used to handle CID deletion when the CID is no longer needed, this is only used by operator. The deletion tracker allows us to implement a `cidDeleteDelay` which is the delay to enqueue another CID event to be reconciled after CID is marked for deletion. This is required for simultaneous CID management by both cilium-operator and cilium-agent. Without the delay, operator might immediately clean up CIDs created by agent, before agent can finish CEP creation. The deletion tracker is not used in cilium-agent and is only used to enforce the delay for deletion by operator.

Related CFP #27752

Draft full implementation: https://github.com/cilium/cilium/pull/33204

### CODE REVIEW

File: operator/pkg/ciliumidentity/cache.go
Comment (asauber) [inline operator/pkg/ciliumidentity/cache.go]: There is an ongoing effort to remove logrus as a dependency from Cilium.

Please switch to using `slog` here.

See:
https://github.com/cilium/cilium/pull/26327
https://github.com/cilium/cilium/pull/33121
https://github.com/cilium/cilium/commit/7511f74c5708891515efbe58d5a6388361e7b8e6

File: operator/pkg/ciliumidentity/cache.go
Comment (asauber) [inline operator/pkg/ciliumidentity/cache.go:321]: Relying on system time for the sake of synchronization is error prone in a distributed system.

The synchronization primitive in Kubernetes is etcd revision numbers via optimistic concurrency. 

My understanding is that there has been some design work on this proposal, is there a CFP that you can link to?

**Review by asauber — CHANGES_REQUESTED** [review]

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go:321]: Hey @asauber, thank you for your review! 

The time is only used locally to know when to schedule the cleanup. We are not actually relying in the exact time. 

CFP https://github.com/cilium/cilium/issues/27752
Draft full implementation: https://github.com/cilium/cilium/pull/33204

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: switched to slog

File: operator/pkg/ciliumidentity/cache.go
Comment (asauber) [reply operator/pkg/ciliumidentity/cache.go:321]: I see, this seems to me to introduce some unspecified (possibly inaccurate, depending on the set system time), duration, during which an ID continues to exist on a Node even when it should not, possibly affecting policy behavior. Is this not the case? (cc @joestringer if you have any additional thoughts)

**Review by asauber — APPROVED** [review]

**Comment by joestringer** [comment]
Could you update the description to describe how this time tracker fits into the design, ie expand the following sentence to explain not just *what* but *how* it integrates into the broader solution?

> This feature will be used by Operator Managing CIDs.

**Comment by ovidiutirla** [comment]
I assume Andrew had similar concern with CIDDeletionTracker, it shouldn't impact the NP enforcement as the NP is done on agent level while the tracker is used to add a configurable delay for CID clean-up. 

Hey @joestringer , thanks for the feedback. Added the following to the PR description:

> `EnqueueTimeTracker` is only be used for metrics to meter the duration from enqueuing the reconciliation until the CID reconciliation is completed. 
`CIDDeletionTracker` is used to handle CID deletion when the CID is no longer needed, this is only used on operator. The deletion tracker allows us to implement a `cidDeleteDelay` which is the delay to enqueue another CID event to be reconciled after CID is marked for deletion. This is required for simultaneous CID management by both cilium-operator and cilium-agent. Without the delay, operator might immediately clean up CIDs created by agent, before agent can finish CEP creation. The deletion tracker is not used in cilium-agent and is only used to enforce the delay for deletion.

**Comment by joestringer** [comment]
@ovidiutirla I still don't quite follow.

> EnqueueTimeTracker is only used for metrics to meter the duration from enqueuing the reconciliation until the CID reconciliation is completed.

Which metrics? What does it mean to reconcile a CID?

File: operator/pkg/ciliumidentity/cache.go
Comment (joestringer) [reply operator/pkg/ciliumidentity/cache.go:321]: I don't think I could speak to the usage of this until there's code that actually uses it. It seems like the reviewers will need to look over the other related PRs to understand this structure before they can know that the usage of the structure doesn't bring up such concerns.

**Comment by asauber** [comment]
In the draft implementation, there is a 30 second delay added to any CID deletion-via-operator here https://github.com/cilium/cilium/pull/33204/files#diff-a5b01710162b5dcb0c820096bd611ded7a00e14a8a60ebc382466ce30d9436caR249-R253

To me, this appears to introduce an [at least] 30 second delay between the operator first processing a delete event and the policy of that delete event being enforced by Agents. I do see the point that even if strong distributed concurrency were in place for CIDs, that would need to be coupled somehow with CEP operations.

**Comment by asauber** [comment]
Apologies for spam. The relevant diff is not linkable on GitHub due to the diff being collapsed by default.

> In the draft implementation, there is a 30 second delay added to any CID deletion-via-operator here https://github.com/cilium/cilium/pull/33204/files#diff-a5b01710162b5dcb0c820096bd611ded7a00e14a8a60ebc382466ce30d9436caR249-R253

https://github.com/cilium/cilium/blob/05e0ebe8b187508fb6fef5e6d6b6fda11d0b6727/operator/pkg/ciliumidentity/reconciler.go#L249-L253

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache.go]: In which case we may end up with a nil `secIds`?

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache.go:180]: What is the benefit of returning a bool instead of an error, here?

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache.go:180]: Why are we removing this helper?

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache.go]: Is it possible to expand this comment explaining why is this required? It would be nice to detail the expected sequence of events that proves the tracker is needed when both operator and agent manage CIDs.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache_test.go]: I'm not sure that a table driven tests approach is the best here. Considering that we only have two subtests and each is slightly different (hence the `setBeforeGet` flag) I suggest to write a simple test without the multiple cases.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache_test.go]: Just as the previous test, it might be better to use two different sets of subtests:

- the first one for the subtests with `unmarkBeforeGet == false`
- the second one for the remaining subtests with `unmarkBeforeGet == true`

This should help in removing some ifs and flags and improving the readability

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks!

I've left some questions around some changes that were previously introduced and reviewed in https://github.com/cilium/cilium/pull/30649.
Also, some suggestions around the structure of the unit tests.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go:180]: This method is not used.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go:180]: Removed for readability. We are only interested if we removed the pod as the error message was static.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Currently, we shouldn't end up with a nil `secIDs`. It was added for clarity.  I removed it if it's obvious that it can not be nil.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Updated the docs

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: refactored the test, splitted by method.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: I feel like it makes more sense for this one to be tested together, though I've renamed the setBeforeGet to increase the readability.

**Review by ovidiutirla — COMMENTED** [review]
Thanks for the review. Refactored a bit the tests to improve the readability. Rebased on main.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: refactored the test, splitted by method.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: I feel like it makes more sense for this one to be tested together, though I've renamed the setBeforeGet to increase the readability.

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [reply operator/pkg/ciliumidentity/cache.go:180]: Would you mind doing that in a separate commit?
I think that all the changes to the `CIDUsageInPods` type (adding the logger, changing this return value and removing the unused `CIDUsedByPod`) should be all part of their own commit, since they have not much to do with the time trackers.

File: operator/pkg/ciliumidentity/cache.go
Comment (pippolo84) [reply operator/pkg/ciliumidentity/cache.go]: I'm sorry but I'm not following the comment.

AFAIU we need the delay (and thus the marking and deletion-tracking) to avoid scenarios where the agent created the CID but it does not have updated the CEP with the CID yet. If the operator sees the new CID and no CEP actually using it, it might be too aggressive and delete the CID. This is supposed to happen more frequently in case of high pod churn (I guess because this increases the delay between the operator receiving the CID creation event and the operator receiving the CEP update event).
Is my understanding correct? Either way, I suggest to rephrase the comment to be more clear and descriptive.

**Comment by sypakine** [comment]
note: I had been commenting on a previous commit version: https://github.com/cilium/cilium/commit/ac7ac99b938035f1d3263c08521c2a27ffd5ff10

I'm not a github expert, so not sure if there is a way to reflect them in this commit...

File: operator/pkg/ciliumidentity/cache_test.go
Comment (pippolo84) [reply operator/pkg/ciliumidentity/cache_test.go]: What I meant was to keep those two cases in the same test, but avoid the table driven tests approach. What do you think about:

```
func TestEnqueueTimeTracker(t *testing.T) {
	tracker := &EnqueueTimeTracker{enqueuedAt: make(map[string]time.Time)}

	// test GetEnqueueTimeAndReset after initial set
	tracker.SetEnqueueTimeIfNotSet("item1")

	_, exists := tracker.GetEnqueueTimeAndReset("item1")
	if !exists {
		t.Error("Expected first GetEnqueueTimeAndReset(\"item1\") to return exists == true, got false")
	}

	_, exists = tracker.GetEnqueueTimeAndReset("item1")
	if exists {
		t.Error("Expected second GetEnqueueTimeAndReset(\"item1\") to return exists == false, got true")
	}

	// test GetEnqueueTimeAndReset without a previous set
	tracker.SetEnqueueTimeIfNotSet("item2")

	_, exists = tracker.GetEnqueueTimeAndReset("item2")
	if !exists {
		t.Error("Expected GetEnqueueTimeAndReset(\"item2\") to return exists == true, got false")
	}
}
```

I think this looks cleaner and also avoid calling two times `GetEnqueueTimeAndReset` in the second case ("without a previous set") where it is not necessary.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the follow-up.
Left some additional comments in the previously opened conversations.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: It address two main concerns:

1. It facilitates CID reuse in scenarios with high pod churn. We are not deleting the CID and we are able to re-use it.
2. Ensure correct CID association in scenarios where the agent rapidly receives create, delete events. 
e.g. we have 1 CID used by one pod, and we churn this pod (agents quickly receives create, delete CID events).  If the agent receives the delete CID, schedules the CEP deletion, then a new create event comes, sees that the CEP is already existing, the scheduled CEP event deletes the CEP.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go:180]: Extracted

**Comment by ovidiutirla** [comment]
> note: I had been commenting on a previous commit version: [ac7ac99](https://github.com/cilium/cilium/commit/ac7ac99b938035f1d3263c08521c2a27ffd5ff10)
> 
> I'm not a github expert, so not sure if there is a way to reflect them in this commit...

Thanks both of you for the feedback! 
I split in a new commit and addressed your changes Mark.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: What do you think about @sypakine proposal with `enqueuedItems` + `expect` pattern? I've updated my implementation to hopefully reflect this.

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: @pippolo84, added more details in the comment. I think once we have the full implementation it will provide enough context.

Comment (pippolo84) [reply operator/pkg/ciliumidentity/cache.go]: Thanks! Not completely sure about the wording yet, but it can be just me. @marseel WDYT? I know you have context from the CfP, any advice? :pray:

File: operator/pkg/ciliumidentity/cache_test.go
Comment (pippolo84) [reply operator/pkg/ciliumidentity/cache_test.go]: Yep, this is better, thanks!

File: operator/pkg/ciliumidentity/cache.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Hey @pippolo84,
Thanks for the feedback, so one concrete example could be during the upgrade, operator might delete an CID that is set locally on CEP without the delayed delete.
Which we could reproduced with namespace label changes, when operator creates a new identity for the new labels, and agent creates a new CID for the same labels, then before operator receives the CEP with the populated CID it thinks this new CID is not used, and it deletes the new CID that was created by agent.

I'm happy to address any other requests to wording even after the further PRs will be merged, do you think we can unblock this and keep the wording discussion offline?

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: I think I kind of understand the intent of it, but I'm not sure if it's required. It seems to be related to https://github.com/cilium/cilium/pull/19936

The current comment in the PR is definitely misleading as it refers to network policy as CEP? 

Regardless, AFAIK Identities are currently only deleted by the operator and the current timeout is 30 minutes + 15 minutes. 
Identitiy is considered alive if any CEP points to CID.
Identitity is considered not live **and** ready for GC if it was not alive for 30 minutes. In this case, identity is marked for deletion using annotation and the next iteration of GC is deleting it if any of the nodes did not remove annotation. 

So essentially (currently) deletion should only happen if:
- Identity was not used by any CEP for 30 minutes (2 * defaults.KVstoreLeaseTTL) - it is marked for deletion at this point
- Identity deletion mark was not removed by any of the nodes for following 15 minutes. (defaults.KVstoreLeaseTTL)

Also, briefly looking at full implementation in https://github.com/cilium/cilium/pull/33204
I don't see any use of it too :thinking:?

Edit: Oh, I see we raced with comments @ovidiutirla :sweat_smile:

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Hey @marseel,  thank you for the feedback, 

The `cidDeletionTracker` is Marked in `handleCIDDeletion` (`handleCIDDeletion` marks or deletes already marked CID)
handleCIDDeletion is called only when, CID only exists in the watcher's store, and it isn't used. 
See https://github.com/cilium/cilium/blob/05e0ebe8b187508fb6fef5e6d6b6fda11d0b6727/operator/pkg/ciliumidentity/reconciler.go

Also `CIDDeletionTracker` was renamed to `DeletionTracker` and `MarkedTime` was renamed to `Get`, I'll update the draft once this is merged.

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: > Which we could reproduced with namespace label changes, when operator creates a new identity for the new labels, and agent creates a new CID for the same labels, then before operator receives the CEP with the populated CID it thinks this new CID is not used, and it deletes the new CID that was created by agent.

I think my comment still applies to your example case. Unless you plan to rework how GC works, I don't think "deletes the new CID that was created by agent." can happen due to timeouts I mentioned in my comment. It would happen only after ~15 minutes by default - as CID was never alive, it would still not be deleted immediately but only marked for deletion first.

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Yes, we plan to rework the GC with operator managing CIDs. The new logic will handle the GC, see https://github.com/cilium/cilium/pull/33381

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: okay, everything makes sense to me. Thanks for explanation :100: 

For the wording part I would take https://github.com/cilium/cilium/pull/19936/files#diff-60a9c671858c5d13fda0333139099edf896a6e3371dc38de7df250a259c6e9a4R113-R125 as an example how to document very similar use-case. "high pod churn" seems unrelated to what it's trying to achieve. With new GC, introducing similar mark to give node a chance to respond during transition period might be worth doing as well.

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: Although, looking at CFP it seems to me like there were changes to GC proposed, but not implementing fully new GC :thinking: ?
https://docs.google.com/document/d/1Hcc_2mB9OOUxrqQgZ-gSYDPnLYE_If_TCzVbUGDOdGM/edit part " CID GC reworked"
It sounds to me that new implementation will be very similar to old one.

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: reworking the existing GC would too much complexity to the existing implementation and branching. Decoupling it and having the GC in the operator managing CIDs adds a clear separation. Also we need to modify the internal state of the CID (operator managing CIDs) which would've added additional complexity in re-working the existing implementation.  But indeed, the new GC should be fairly similar to the existing one.

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: I also consider the delayed deletion is useful in high pod churn scenarios as we do not have to delete and re-create the same CID over and over..

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: Thanks for all the explanation. I will defer to @joestringer if implementing a new GC in this case makes sense or not. 

My gut feeling is that implementing a separate GC will make it actually more complex - especially during migration from "current state" to "operator managing CIDs". GC in operator seems to me totally unrelated to who is managing CIDs.
CFP part of "CID GC reworked" sounds like the implementation of this could happen even if operator does not manage CIDs by introducing this `DeletionTracker` and replacing Heartbeat annotation with it.

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: > I also consider the delayed deletion is useful in high pod churn scenarios as we do not have to delete and re-create the same CID over and over..

sure, but that's already implemented in the current GC.

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go]: Updated the docs.

@dlapcevic do you have any thoughts around the existing GC and the one proposed to be in operator managing CIDs?

Comment (marseel) [reply operator/pkg/ciliumidentity/cache.go]: I will try to sum it up and let me know if my understanding is correct so @joestringer and @dlapcevic  doesn't have to read all of that conversation :sweat_smile: 

We are considering how GC behaves in two different cases:
(1) During the transition period from cilium-agent creating CIDs to operator creating CIDs when both operator and agent can create CIDs
(2) How GC behaves once operators is fully managing CIDs

Let's think about the current GC and how it would behave:
(1) if CID was ever used, it will be marked for deletion after 30 minutes of not being used with annotation and deleted after another 15 minutes - this already addresses your concern with high-churn pods. 
If CID was never used by any CEP, it will mark it for deletion and delete it after 15 minutes.
Sounds reasonable both for identities created by the operator and the agent.
CIDs created by the agent and not yet used by any CEPs won't be deleted immediately. 
 
(2) It will still work the same, but adding an annotation mark for deletion won't make much sense as the agent won't be able to remove it anymore.
So proposed switch from `deletion mark annotation` to `DeletionTracker` is a nice optimization in terms of API calls, but not necessary for it to work correctly. It should be fairly easy to implement it in the current GC though.  


It could probably even be instantly enabled ( `deletion mark annotation` -> `DeletionTracker`) when switching from agent creating CIDs to operator creating CIDs during migration and the current GC would still behave correctly in both cases (1) and (2), but would not react to deletion mark annotation, but used fixed delay for waiting on CEP.

Of course interval/timeouts (30/15 minutes) can be already changed IIRC, I was just using defaults as examples. 


So the question is, what would be the purpose of implementing a new GC, instead of implementing this improvement to the current GC, which seems significantly easier?

File: operator/pkg/ciliumidentity/cache_test.go
Comment (sypakine) [inline operator/pkg/ciliumidentity/cache_test.go]: - name field is not in-line with the rest
- provide the parameter fields for test structs

```
{
    name:                  "set_and_get",
    enqueuedItems: []string{"item"},
    expectItem:         "item",
    expectExists:       "true",
}
```

This way, each test case is self-documenting.

Google-internal: see go tips 21 and 50.

File: operator/pkg/ciliumidentity/cache.go
Comment (dlapcevic) [reply operator/pkg/ciliumidentity/cache.go]: In my opinion, GC is part of CID management, and there is no reason for it to be separated from the CID controller. The CID controller anyway needs to know the desired state of CIDs, when CIDs are used and when not, and adding deletion to it is not a complex effort.

Beside that I see a few benefits:
1. Performance and scalability\
a) Smaller number of calls to kube-apiserver without adding and removing the mark. Remember: every update needs to be sent to all nodes.
b) Processing all CIDs at one time (each GC run) instead of continuously, creates a spike of requests that can significantly affect KCP and cilium-operator's k8s client’s rate limiting.
2. Ease of use\
No need to configure GC.
3. Security\
Remove write permission for CIDs from cilium-agent.\
When cilium-agent is not managing CIDs, it shouldn’t update CIDs, so it won’t need write permission to CIDs.

We had issues related to these points.
We also had issues where cilium-operator was having connection instability with kube-apiserver so it was restarting every 15-20 minutes (or leader was changed), and in this case we ended up having no CIDs cleaned up and eventually hitting the 65k CID limit.

File: operator/pkg/ciliumidentity/cache_test.go
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go:340]: What do you mean by `provide the parameter fields for test structs`? Adding a new bool to the test case struct for the second call to `GetAndReset`?

**Review by dlapcevic — APPROVED** [review]
lgtm

File: operator/pkg/ciliumidentity/cache.go
Comment (joestringer) [reply operator/pkg/ciliumidentity/cache.go]: I think that this topic would be a lot more approachable with a more incremental approach. Use the current GC first, propose improvements to the current GC implementation and push on that independently. Then propose the PR that refactors/removes the current GC implementation to merge it into this CID controller. Each of those steps would be self-contained, provide value, and could be individually assessed for their benefits and drawbacks.

If we were looking at a PR that was titled "Improve cilium-operator Identity garbage collection" with the PR description containing the text in the [post immediately above](https://github.com/cilium/cilium/pull/33380#discussion_r1666851432), then I think we would be in a much better position to critique the design and implementation, and debate how it affects the implementation.

Given how critical these operations are and how nuanced they can be, I don't think we're interested in having multiple implementations in the tree at the same time. Not unless there's something critical to the design that I'm missing which makes the garbage collection improvements inherently different.

EDIT: Let me adjust a bit on multiple GC: If we want to trial and incrementally roll out two implementations, then it could be useful to have a swappable implementation where you define which GC algorithm to use based on a flag. But for that, ideally we would just structure it so that the GC aspect fits into the broader code in a consistent way and the only difference is the details of the GC which are swapped in/out based on a flag. If that's what you're going for, I think that at least provides us a path from the current implementation to allowing the new implementation, enabling by default, deprecating the old one, then removing the old one. However if we have sufficient testing I don't think we necessarily need to go through all of that, we could just stick to one implementation that is incrementally improved/changed.

**Comment by ovidiutirla** [comment]
Closing this in favor of:
* CID Event Tracker added as part of the metrics https://github.com/cilium/cilium/pull/34128
* CID Deletion Tracker removed as part of the feedback from https://github.com/cilium/cilium/pull/33380#discussion_r1669494701 the existing identitygc will be used.

**Pull request closed without merging**
